### PR TITLE
Resample and augmentation: the separable walk, bounds once, ITK's input reused

### PR DESCRIPTION
**Pull request by vboussot**
## Summary

Resample's walk and the augmentation draws stop recomputing and copying what they already hold. A stored translation or axis-aligned scale (an `AffineStage` whose folded matrix is exactly diagonal) now takes the separable walk instead of the per-voxel float64 walk, applied per component as the walk applies it; the separable and general gathers pick their taps by an advanced index instead of an inner-axis `index_select`; the slabbed general path lands each gathered slab in one preallocated output instead of a `torch.cat`; `walk_rows` probes its default budget once a second per device instead of once a patch, and `Resample._sample` takes a budget handed down; `DisplacementStage.bound_xyz` becomes a `cached_property` dropped from `__getstate__`; the plan records per case what the pulls need (`_StoredMap`: the folded bound, whether the map is its affine part) and the run decodes a case's stages where it samples them, keeping the last cases' within 512 MiB and dropping a case's with its `stream_abort`; the host route holds one ITK input image and fills it in place from one region to the next through `_SetImageFromArray` (SimpleITK/SimpleITK#2683), replaced on a change of shape or dtype, dropped by the whole-volume call and `stream_abort`, never pickled. On the augmentation side, a copy's records re-fold when the copy changes rather than before every patch; `_load_augmentation_group` hands the draws the case itself rather than a clone per copy (`Foreign` alone clones); `Elastix` keeps only the float32 sampling grid and `reset_state` drops it with the draw; `EulerTransform._source_coordinates` runs every step past the affine map in place on the block's device (float64 on a device so a TF32 setting does not reach it; the host path bitwise the chain of temporaries it replaced) and skips the reflection and clip when the mapped box lies inside `[margin, n - 1 - margin]`, margin the larger of 1e-3 voxel and 1e-5 of the extent; `CutOUT` broadcasts its `[1, *spatial]` mask over the channels. The draw contract (`_compute` returns a fresh tensor or a view and never writes into what it is handed: the copies alias the case) is stated in AGENTS.md and the base class; docstrings state what the epoch report's device phases are, what a window widened mid-run holds, the read schedule's own count and the chunk cache's floor.

## Measured

| Change | Command | Before | After |
|---|---|---|---|
| Diagonal stored map on the separable walk, 256^3 float32 region through a translation | `.scratch/bench/bench_separable.py` | general walk 2.4-3.6 s on the host, 0.04 s on CUDA | 0.13-0.15 s (ITK's host resampler 0.11-0.14 s), 0.003 s on CUDA |
| Taps by an advanced index, `[1, 256, 256, 256]` float32 on the host | `.scratch/micro_take.py`, `.scratch/micro_flat.py` | 0.5-1.5 s per inner-axis `index_select`; flat pick 125 ms (one channel), 477 ms (four) | 9 ms; 27 ms, 171 ms; 0.3 ms and 2 ms either way on CUDA |
| Separable resample of a 256^3 region on the host | `.scratch/bench/bench_separable.py cpu` | 8.3-9.9 s (the general walk 3.6 s) | 0.13-0.15 s |
| `walk_rows` default budget, 1001 calls | `.scratch/bench/bench_walk_rows.py` | 197 us (cpu), 211 us (cuda) per call, one probe per call | 3 us (cpu), 4 us (cuda), one probe in 1001 calls |
| Slabs landed in the output, CUDA, 4-row slabs over 256^3 float32 through a rotation, 64 MiB output | `.scratch/bench/bench_cat.py 256,256,256 4` | 128 MiB above resident | 94 MiB; 46 vs 48 MiB on 96x192x192 |
| Stored map kept as its bound, 10 cases of a 40x64x64 dense field | `.scratch/bench/bench_stored.py` | 37.5 MiB pickled, 10 entries resident after the plan and 10 after sampling one case | 6 KB, 0 after the plan, 1 after sampling a case |
| `DisplacementStage.bound_xyz`, 3x160x256x256 float64 field, 5 calls | `.scratch/bench/bench_bound.py` | 72.6-74.5 ms per call | 73.7 ms the first call, 0.0 ms the next four |
| ITK input filled in place, SimpleITK 2.5.5, 384^3 uint16 (113 MB), 11 repetitions, OMP_NUM_THREADS=4 | `.audit-local/bench/sitk_outputimage_gist.py 11` | `GetImageFromArray` min 44.3 ms, median 48.0 ms | min 7.5 ms, median 8.6 ms; min 6.1 ms, median 8.9 ms with a `sitk.Resample` executed in between |
| ExaSPIM, 513x1331x1776 uint16, 40 blocks, cpu exact, two runs each | `.audit-local/bench/bench_tree.sh` | 5.6 / 5.7 s | 5.2 / 5.2 s; 0 differing voxels of 2,363,856 per z slice (0, 256, 512) |
| Re-fold on a change of copy, one case of 3 copies through Permute and Resample, 36 patches | `.scratch/bench/bench_refold.py` | `_fold_case_state` called 72 times | 6 calls |
| Group handed the case, 10 copies of a 64 MiB float32 case under Flip | `.scratch/bench/bench_augment.py` | 10 clones, 0.22 s, 640 MiB of memcpy per group per case | 0 clones, 0.15 s |
| Elastix grids, 4 copies of a 64^3 case | `.scratch/bench/bench_elastix.py` | 24.0 MiB of float64 displacements beside 12.0 MiB of grids, grids surviving `reset_state` | 0 MiB beside the 12.0 MiB, 0 grids after `reset_state` |
| Euler coordinates, 192^3 region under a rotation | `.scratch/bench/bench_euler.py` | 0.43-0.49 s and 867 MiB of ru_maxrss growth for an 81 MiB result on the host | 0.32-0.39 s and 355 MiB on the host; 0.01 s on CUDA with 265 MiB of device peak |
| CutOUT mask broadcast, 8x128^3 float32 | `.scratch/bench/bench_cutout.py` | 54-60 ms per apply, one repeat | 20-40 ms, no repeat |

**Values.** Two routes move, at rounding. A diagonal stored map now goes through the separable gather, measured against the two routes it replaces on the 22x28x34 phantom (`.scratch/measure_sep.py`): a linear float32 blend by at most 4.5e-4 (1.3e-6 of the range) against the general walk's `grid_sample` and 1.5e-5 (4.3e-8, one float32 ulp) against the host resampler; cubic 7.6e-5 (2.1e-7); an int16 linear blend by one unit on at most 47 of 20944 voxels (walk) and 21 (host); a nearest pick by nothing on either, and byte-identical to `sitk.Resample` on a label map. A CUDA block's Euler coordinates come from a device matmul where they were computed on the host and copied: measured with the float32 device matmul over 20 random rotations at 64x72x80 (`.scratch/measure_euler.py`), at most 1.53e-5 voxel apart (one float32 ulp), 13.7% of the entries in the last bit; the map now runs in float64 on a device and lands in float32, and the host path is bitwise the chain of temporaries it replaced. Every other change: values do not move (bit-identical, or `torch.equal` against the spelling it replaces).

**Tests.** `tests/unit/test_resample_transform.py`: the two oracle tests for the diagonal map (within the routes it replaces; byte-identical to SimpleITK on a label map), the plan keeps a case's bound while the run holds one case's stages, each slab landed in the one output, one shape fills the ITK image in place while another dtype or shape replaces it, a sweep keeps the image across regions while the pickled state, `stream_abort` and a whole-volume call drop it. `test_sampling_slabs.py`: the default budget probed once per second, `torch.equal` against the general walk for a translation, a scale and their compositions with flipped axes, a rotation and a displacement stage refused. `test_geometry.py`: the bound is one pass per stage and stays out of the pickle. `test_augmentation.py`: Euler coordinates bitwise the chain of temporaries over random rotations (2D and 3D, interior and border-crossing regions, a singleton axis) and built on the block's device, Elastix one grid per copy dropped with the draw, CutOUT broadcast pinned against the repeated-and-tested spelling, a group handed the case, not a clone. `test_case_expansion.py`: a copy's records re-folded once per change of copy. `test_cohort_walk.py`, `test_sweep_pipeline.py`, `test_transformer_workflow.py`: a subset asked for its names once per walk, the resident-block formula, the snapshot's depth.

**Stack.** Stacked on `pr/1.8.2-06-training`; merge after it. Before deleting this branch, retarget the next PR of the stack.

## Summary by CodeRabbit

* **New Features**
  * Improved streamed resampling performance through memory reuse, bounded caching, and direct slab output assembly.
  * Added optimized handling for diagonal affine transformations and device-aware sampling.
  * Improved augmentation safety by preventing unintended input mutation and reducing unnecessary memory copies.
  * Included decoded OME-Zarr chunks in pipeline memory budgeting.

* **Bug Fixes**
  * Corrected augmentation copy behavior, transform-state refolding, and cache reset handling.
  * Improved memory lifecycle management for resampling and external transforms.

* **Tests**
  * Added extensive coverage for transformations, sampling, caching, memory behavior, and copy isolation.

### CODE REVIEW

**Comment by vboussot** [comment]
@coderabbitai review

File: konfai/data/transform.py [hunk 1]
```diff
@@ -32,8 +33,10 @@
 
 try:
     import SimpleITK as sitk
+    from SimpleITK.SimpleITK import _SetImageFromArray as _set_image_from_array
 except ImportError:
     sitk = None  # type: ignore[assignment]
+    _set_image_from_array = None
```
[the remark this answers is not in the document]
Comment (vboussot) [reply konfai/data/transform.py:39]: Fixed, and it was worse than reported: a SimpleITK that does not export that private symbol made `sitk = None`, so the run said "install SimpleITK" for an installed SimpleITK. Split into `_optional_image_filler()`; the caller already handled `None`.

File: konfai/network/network.py [hunk 2]
```diff
@@ -301,7 +301,8 @@ def recorded(self) -> int:
 
         def set_window(self, n: int) -> None:
             """Keep at least the last ``n`` values and weights. Grows only; until it is called the
-            history is unbounded."""
+            history is unbounded. A window widened mid-run holds what it had until ``n`` more values
+            arrive: the reads in between average the values held, fewer than they ask for."""
```
[the remark this answers is not in the document]
Comment (vboussot) [reply konfai/network/network.py:305]: Fixed: a widened window refills `n - m`, not `n`.

File: tests/unit/test_augmentation.py [hunk 3]
```diff
@@ -456,3 +456,148 @@ def counting_read_image(path: str):
...
+    flip = Flip(f_prob=[1.0, 1.0, 1.0])
+    flip.load(0.5)
+    group = DataAugmentationsList(nb=8)
+    group.data_augmentations = [flip]
+    manager = DatasetManager(0, "CT", "CT", "CASE", dataset, None, [], [group])
+    manager._load([])
+    before = manager.data[0].clone()
+    manager._load_augmentation_group(1, group)
+    selected = flip.who_index[0]
+    assert 0 < len(selected) < 8, "the fixture wants selected and unselected copies"
```
[the remark this answers is not in the document]
Comment (vboussot) [reply tests/unit/test_augmentation.py:595]: Fixed. It was a real 1-in-128 flake (`torch.rand(8) < 0.5`); the selection is seeded locally and the premise assertion stays as the guard.
